app/models/exercise_map.rb

```diff
@@ -23,7 +23,10 @@ def map
   def exercises
     @exercises ||= track.
       exercises.
-      map { |e| ExerciseWithSolution.new(e, solution_for(e)) }
+      left_outer_joins(:solutions).
+      includes(:unlocks).
+      where("solutions.user_id = ? OR solutions.user_id IS NULL", user.id).
+      map { |e| ExerciseWithSolution.new(e) }
```

Comment:
In line with my comment below, I think this should be `ExerciseWithSolution.new(e, e.solutions.first)`